massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+        let config = self.grpc_config.clone();
+        let storage = self.storage.clone_without_refs();
+
+        let (tx, rx) = tokio::sync::mpsc::channel(config.max_channel_size);
+        let mut in_stream = request.into_inner();
+
+        tokio::spawn(async move {
+            while let Some(result) = in_stream.next().await {
+                match result {
+                    Ok(req_content) => {
+                        if req_content.endorsements.is_empty() {
+                            let _res = send_endorsements_notify_error(
+                                req_content.id.clone(),
+                                tx.clone(),
+                                tonic::Code::InvalidArgument,
+                                "the request payload is empty".to_owned(),
+                            )
+                            .await;
+                        } else {
+                            let proto_endorsement = req_content.endorsements;
+                            if proto_endorsement.len() as u32 > config.max_endorsements_per_message
+                            {
+                                let _res = send_endorsements_notify_error(
+                                    req_content.id.clone(),
+                                    tx.clone(),
+                                    tonic::Code::InvalidArgument,
+                                    "too many endorsements".to_owned(),
+                                )
+                                .await;
+                            } else {
+                                let endorsement_deserializer =
+                                    SecureShareDeserializer::new(EndorsementDeserializer::new(
+                                        config.thread_count,
+                                        config.endorsement_count,
+                                    ));
+                                let verified_eds_res: Result<HashMap<String, SecureShareEndorsement>, ModelsError> = proto_endorsement
+                                .into_iter()
+                                .map(|proto_endorsement| {
+                                    let mut ed_serialized = Vec::new();
+                                    ed_serialized.extend(proto_endorsement.signature.as_bytes());
+                                    ed_serialized.extend(proto_endorsement.creator_public_key.as_bytes());
+                                    ed_serialized.extend(proto_endorsement.serialized_content);
+                                    let verified_op = match endorsement_deserializer.deserialize::<DeserializeError>(&ed_serialized) {
+                                        Ok(tuple) => {
+                                            let (rest, res_endorsement): (&[u8], SecureShareEndorsement) = tuple;
+                                                if rest.is_empty() {
+                                                if let Ok(_verify_signature) = res_endorsement.verify_signature() {
```

Comment:
todo : put it in function for remove duplicate code

Reply:
we'll have 3 signatures verification ? Endorsement, Block and Operation... or we will did an all in one ?

Reply:
i need to look more at it

Reply:
KISS, we'll do an optimisation phase before merging